samples/subsys/usb/uac2_explicit_feedback/src/main.c

```diff
@@ -0,0 +1,683 @@
+	 * SOF n appears on I2S during SOF n+3, the Ziegler Nichols Ultimate
+	 * Gain is around 1.15 and the oscillation period is around 90 SOF.
+	 * (much nicer oscillations with 204.8 SOF period can be observed with
+	 * gain 0.5 when the delay is not n+3, but n+33 - surprisingly the
+	 * resulting PI coefficients after power of two rounding are the same).
+	 *
+	 * Ziegler-Nichols rule with applied stability margin of 2 results in:
+	 *   Kc = 0.22 * Ku = 0.22 * 1.15 = 0.253
+	 *   Ti = 0.83 * tu = 0.83 * 80 = 66.4
+	 *
+	 * Converting the rules above to parallel PI gives:
+	 *   Kp = Kc = 0.253
+	 *   Ki = Kc/Ti = 0.254/66.4 ~= 0.0038253
+	 *
+	 * Because we want fixed-point optimized non-tunable implementation,
+	 * the parameters can be conveniently expressed with power of two:
+	 *   Kp ~= pow(2, -2) = 0.25    (divide by 4)
+	 *   Ki ~= pow(2, -8) = 0.0039  (divide by 256)
+	 *
+	 * This can be implemented as:
+	 *   ctx->integrator += error;
+	 *   return (error + (ctx->integrator / 64)) / 4;
+	 * but unfortunately such regulator is pretty aggressive and keeps
+	 * oscillating rather quickly around the setpoint (within +-1 sample).
+	 *
+	 * Manually tweaking the constants so the regulator output is shifted
+	 * down by 4 bits (i.e. change /64 to /2048 and /4 to /128) yields
+	 * really good results (the outcome is similar, even slightly better,
+	 * than using I2S LRCLK edge counting directly).
+	 */
+	ctx->integrator += error;
+	return (error + (ctx->integrator / 2048)) / 128;
+}
+
+static void process_feedback(struct usb_i2s_ctx *ctx)
+{
+	uint32_t sof_cc;
+	uint32_t framestart_cc;
+	uint32_t fb;
+
+	sof_cc = nrfx_timer_capture_get(&feedback_timer_instance,
+		FEEDBACK_TIMER_USBD_SOF_CAPTURE);
+	framestart_cc = nrfx_timer_capture_get(&feedback_timer_instance,
+		FEEDBACK_TIMER_I2S_FRAMESTART_CAPTURE);
+
+	update_sof_offset(ctx, sof_cc, framestart_cc);
+
+	if (IS_ENABLED(CONFIG_USE_I2S_LRCLK_EDGES_COUNTER)) {
+		int32_t offset = ctx->rel_sof_offset + ctx->base_sof_offset;
+
+		ctx->fb_counter += sof_cc;
+		ctx->fb_periods++;
+
+		if (ctx->fb_periods == BIT(FEEDBACK_K - FEEDBACK_P)) {
+
+			/* fb_counter holds Q10.10 value, left-justify it */
+			fb = ctx->fb_counter << FEEDBACK_FS_SHIFT;
+
+			/* Align I2S FRAMESTART to USB SOF by adjusting reported
+			 * feedback value. This is endpoint specific correction
+			 * mentioned but not specified in USB 2.0 Specification.
+			 */
+			if (abs(offset) > BIT(FEEDBACK_P)) {
+				fb += offset_to_correction(offset);
+			}
+
+			ctx->fb_value = fb;
+			ctx->fb_counter = 0;
+			ctx->fb_periods = 0;
+		}
+	} else {
+		/* Use PI controller to generate required feedback deviation
+		 * from nominal feedback value.
+		 */
+		fb = SAMPLES_PER_SOF << (FEEDBACK_K + FEEDBACK_FS_SHIFT);
+		/* Clear the additional LSB bits in feedback value, i.e. do not
+		 * use the optional extra resolution.
+		 */
+		fb += pi_update(ctx) & ~0xF;
+		ctx->fb_value = fb;
+	}
+}
+#else
```

Comment:
I have been looking at this for a long time and have not finished it yet. But yes, the most problematic thing here is this sample. It should be possible to move this block to a separate file and required defines to a common header. Also, it would be nice if the example would work without the I2S controller, even if it makes no functional sense. As it is, it cannot go in.

Reply:
While moving this to separate file might be somewhat possible (I would prefer to move the sample to `samples/boards/nrf` instead and just remove the ifdefs), making it work without I2S controller is absolute nonsense. What could be done is to replace I2S with some other audio sink, but that doesn't really bring much additional value over the sample with I2S-only (the other audio sink would still need platform-specific code for timestamping). The absolute crucial part is that there must be some audio sink with its own free-running clock. Removing the audio sink is essentially making the external feedback part completely irrelevant (and the sample has external feedback in its name!).